File: src/main/java/org/cubexmc/metro/model/RoutePoint.java
package org.cubexmc.metro.model;

import org.bukkit.Location;

public record RoutePoint(String worldName, double x, double y, double z) {

    public static RoutePoint fromLocation(Location location) {
        if (location == null || location.getWorld() == null) {
            return null;
        }
        return new RoutePoint(location.getWorld().getName(), location.getX(), location.getY(), location.getZ());
    }

    public static RoutePoint fromConfigString(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }

        String[] parts = value.split(",");
        if (parts.length != 4) {
            return null;
        }

        try {
            return new RoutePoint(
                    parts[0],
                    Double.parseDouble(parts[1]),
                    Double.parseDouble(parts[2]),
                    Double.parseDouble(parts[3])
            );
        } catch (NumberFormatException ignored) {
            return null;
        }
    }

    public String toConfigString() {
        return worldName + "," + x + "," + y + "," + z;
    }

    public double distanceSquared(RoutePoint other) {
        if (other == null || !worldName.equals(other.worldName())) {
            return Double.MAX_VALUE;
        }
        double dx = x - other.x();
        double dy = y - other.y();
        double dz = z - other.z();
        return dx * dx + dy * dy + dz * dz;
    }
}

File: src/main/java/org/cubexmc/metro/model/RoutePoint.kt
package org.cubexmc.metro.model

import org.bukkit.Location

data class RoutePoint(
    private val worldName: String,
    private val x: Double,
    private val y: Double,
    private val z: Double,
) {
    fun worldName(): String = worldName

    fun x(): Double = x

    fun y(): Double = y

    fun z(): Double = z

    fun toConfigString(): String = "$worldName,$x,$y,$z"

    fun distanceSquared(other: RoutePoint?): Double {
        if (other == null || worldName != other.worldName()) {
            return Double.MAX_VALUE
        }
        val dx = x - other.x()
        val dy = y - other.y()
        val dz = z - other.z()
        return dx * dx + dy * dy + dz * dz
    }

    companion object {
        @JvmStatic
        fun fromLocation(location: Location?): RoutePoint? {
            val world = location?.world ?: return null
            return RoutePoint(world.name, location.x, location.y, location.z)
        }

        @JvmStatic
        fun fromConfigString(value: String?): RoutePoint? {
            if (value.isNullOrBlank()) {
                return null
            }
            val parts = value.split(",")
            if (parts.size != 4) {
                return null
            }
            return try {
                RoutePoint(
                    parts[0],
                    parts[1].toDouble(),
                    parts[2].toDouble(),
                    parts[3].toDouble(),
                )
            } catch (_: NumberFormatException) {
                null
            }
        }
    }
}
